specification/flag-evaluation/flag-evaluation.md

```diff
@@ -0,0 +1,195 @@
+
+##### Requirement 1.10
+
+> The `client` **MUST** provide methods for flag value evaluation with the value wrapped in a `evaluation details` object, with parameters `flag key` (string, required), `default value` (boolean | number | string | object, required), `evaluation context` (optional), and `evaluation options` (optional).
+
+```
+// example detailed boolean flag evaluation
+FlagEvaluationDetails<boolean> myBoolDetails = client.getBooleanDetails('bool-flag', false);
+
+// example detailed string flag evaluation
+FlagEvaluationDetails<string> myStringDetails = client.getStringDetails('string-flag', 'N/A', evaluationContext, options);
+
+```
+
+##### Requirement 1.11
+
+> The `evaluation details` wrapper **MUST** have a `value` field matching the type of the flag value, and containing to the flag value.
+
+##### Condition 1.12
+
+> The language supports generics (or an equivalent feature).
+>
+> ##### Conditional Requirement 1.12.1
+>
+> > The `evaluation details` wrapper **SHOULD** accept a generic argument (or utilize an equivalent language feature) which indicates the type of the wrapped `value` field.
+
+##### Requirement 1.13
+
+> The `evaluation details` wrapper **MUST** have a `flag key` field field of type string, representing the flag key.
+
+##### Requirement 1.14
+
+> The `evaluation details` wrapper **MUST** have an optional `variant` field of type string which represents a space-constrained identifier for a particular flag value.
+
+##### Requirement 1.15
+
+> The `evaluation details` wrapper **MUST** have a `reason` field of type string indicating the semantic reason for the returned flag value. Possible values vary by provider, but may include `"TARGETING_MATCH"`, `"SPLIT"`, `"DISABLED"`, `"DEFAULT"`, `"UNKNOWN"` and `"ERROR"`.
+
+##### Requirement 1.16
+
+> In cases of abnormal execution (network failure, unhandled error, etc) the `reason` field in the `evaluation details` **MUST** be set to `"ERROR"`.
+
+##### Requirement 1.17
+
+> The `evaluation details` wrapper **MUST** have an optional `error code` field of type string, which, in cases of abnormal execution, identifies an error occurred during flag evaluation, with possible values `"PROVIDER_NOT_READY"`, `"FLAG_NOT_FOUND"`, `"PARSE_ERROR"`, `"TYPE_MISMATCH"`, or `"GENERAL"`.
+
+##### Requirement 1.18
+
+> If the `reason` in the `evaluation details` is set to `"ERROR"`, the sibling `error code` field **MUST** be set.
+
+##### Requirement 1.19
+
+> The `evaluation details` wrapper **MUST** have a `executed hooks` field, containing an object which itself contains 4 arrays, identified with keys (`before`, `after`, `error` and `finally`) indicating the hooks run for the respective life-cycle stages.
+
+```
+executedHooks: {
+  before: [],
+  after: [],
+  error: [],
+  finally: []
```

Comment:
Not sure how we want to handle it, but 'finally' is a java reserved word. I've chosen "afterAll" instead.

Reply:
Oh, right. I like `finally` because really these sort of hooks behave just like a `finally` block, so it's intuitive. But if it's a reserved word in Java, it may be in other languages as well (frankly I'm a bit surprised the TS parser has no issue with it).

I'm OK with making it "afterAll" in general, or we could say it's `finally` OR `afterAll`.

Reply:
Avoiding using well known reserved words is a good idea, but in general it may be impossible to completely avoid reserved words. Maybe add some guidance to the spec that if a reserved word is used then choose an appropriate substitute.

Reply:
We're going to remove this for now anyway, as discussed in other comments, unless anyone disagrees. @justinabrahms @dyladan